### ISSUE TITLE
Misleading error message with negative range

### ISSUE THREAD

**Issue by vultix** [rust-lang/rust#83413]
Given the following code: [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=046e611931de232ad60923dac82f78b0)

```rust
fn main() {
  for x in -5..5 {
      do_something(x);
  }    
}

fn do_something(val: usize) {}
```

The current output is:

```
error[E0277]: the trait bound `usize: Neg` is not satisfied
 --> src/main.rs:2:12
  |
2 |   for x in -5..5 {
  |            ^^ the trait `Neg` is not implemented for `usize`

error: aborting due to previous error
```

This created confusion for a new Rust user on our team.

Ideally the range type would be inferred to be `i32`:

```
error[E0308]: mismatched types
 --> src/main.rs:3:20
  |
3 |       do_something(x);
  |                    ^ expected `usize`, found `i32`
  |
help: you can convert an `i32` to a `usize` and panic if the converted value doesn't fit
  |
3 |       do_something(x.try_into().unwrap());
  |                    ^^^^^^^^^^^^^^^^^^^^^
```

_Alternatively_ the output could mention why the type was inferred to be usize:

```
  |
2 |   for x in -5..5 {
  |            ^^ the trait `Neg` is not implemented for `usize`
  |
help: type `usize` was inferred due to call on line 3
  |
3 |      do_something(x);
  |                   ^
```

**Comment by leonardo-m** [comment]
Wow, explanations for problems caused by type inference, nice :)

**Comment by vvzen** [comment]
As a new starter on Rust, this also caught me off guard. Thanks for the explanation! Was there any progress made here ?

**Comment by estebank** [comment]
Current output:

```
error[E0277]: the trait bound `usize: Neg` is not satisfied
 --> src/main.rs:2:12
  |
2 |   for x in -5..5 {
  |            ^^ the trait `Neg` is not implemented for `usize`
  |
  = help: the following other types implement trait `Neg`:
            isize
            i8
            i16
            i32
            i64
            i128
            f32
            f64
          and 8 others
```

It should likely suggest writing `for x in -5i32..5 {` instead (which will cause subsequent type errors in the fn call).
